README.md
[![Build Status](https://travis-ci.org/moznion/Regexp-Lexer.svg?branch=master)](https://travis-ci.org/moznion/Regexp-Lexer)
# NAME

Regexp::Lexer - Lexer for regular expression of perl

# SYNOPSIS

    use Regexp::Lexer qw(tokenize);
    my $tokens = tokenize(qr{\Ahello\s+world\z}i);

# DESCRIPTION

Regexp::Lexer is a lexer for regular expression of perl.

This module splits the regular expression string to tokens
which has minimum meaning.

# FUNCTIONS

- `tokenize($re:Regexp)`

    Tokenizes the regular expression.

    This function takes a argument as `Regexp`, namely it must be regexp quoted variable (i.e. `qr/SOMETHING/`).
    If not `Regexp` argument is given, this function throws exception.
    And invalid syntax regular expression is given, also this function throws exception.

    This function returns the result like so;

        {
            tokens => [
                {
                    char => '\A',
                    index => 1,
                    type => {
                        id => 67,
                        name => 'EscapedBeginningOfString',
                    },
                },
                {
                    char => 'h',
                    index => 2,
                    type => {
                        id => 1,
                        name => 'Character',
                    },
                },
                ...
            ],
            modifiers => ['^', 'i'],
        }

    `tokens` is the token list. Information `type` of token is located in the [Regexp::Lexer::TokenType](https://metacpan.org/pod/Regexp::Lexer::TokenType).

    `modifiers` is the list of modifiers of regular expression. Please see also [perlre](https://metacpan.org/pod/perlre).

# SEE ALSO

- [perlre](https://metacpan.org/pod/perlre)
- [perlrebackslash](https://metacpan.org/pod/perlrebackslash)
- [Regexp::Lexer::TokenType](https://metacpan.org/pod/Regexp::Lexer::TokenType)

# LICENSE

Copyright (C) moznion.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

moznion <moznion@gmail.com>


Changes
Revision history for Perl extension Regexp-Lexer

{{$NEXT}}

0.05 2015-02-07T06:53:44Z

    - Suppress warnings which occur on comparing undefined characters

0.04 2014-11-19T03:06:07Z

    - Discontinue supporting perl-5.8 or earlier

0.03 2014-11-17T15:40:53Z

    - Fix tests
    - Fix documentations

0.02 2014-11-16T15:23:33Z

    - Fix handling for not escaped double quote

0.01 2014-11-16T06:02:33Z

    - original version



lib/Regexp/Lexer.pm
package Regexp::Lexer;
use 5.010001;
use strict;
use warnings;
use B;
use Carp qw/croak/;
use Regexp::Lexer::TokenType;
use parent qw(Exporter);

our @EXPORT_OK = qw(tokenize);

our $VERSION = "0.04";

my %escapedSpecialChar = (
    t => Regexp::Lexer::TokenType::EscapedTab,
    n => Regexp::Lexer::TokenType::EscapedNewline,
    r => Regexp::Lexer::TokenType::EscapedReturn,
    f => Regexp::Lexer::TokenType::EscapedFormFeed,
    F => Regexp::Lexer::TokenType::EscapedFoldcase,
    a => Regexp::Lexer::TokenType::EscapedAlarm,
    e => Regexp::Lexer::TokenType::EscapedEscape,
    c => Regexp::Lexer::TokenType::EscapedControlChar,
    x => Regexp::Lexer::TokenType::EscapedCharHex,
    o => Regexp::Lexer::TokenType::EscapedCharOct,
    0 => Regexp::Lexer::TokenType::EscapedCharOct,
    l => Regexp::Lexer::TokenType::EscapedLowerNext,
    u => Regexp::Lexer::TokenType::EscapedUpperNext,
    L => Regexp::Lexer::TokenType::EscapedLowerUntil,
    U => Regexp::Lexer::TokenType::EscapedUpperUntil,
    Q => Regexp::Lexer::TokenType::EscapedQuoteMetaUntil,
    E => Regexp::Lexer::TokenType::EscapedEnd,
    w => Regexp::Lexer::TokenType::EscapedWordChar,
    W => Regexp::Lexer::TokenType::EscapedNotWordChar,
    s => Regexp::Lexer::TokenType::EscapedWhiteSpaceChar,
    S => Regexp::Lexer::TokenType::EscapedNotWhiteSpaceChar,
    d => Regexp::Lexer::TokenType::EscapedDigitChar,
    D => Regexp::Lexer::TokenType::EscapedNotDigitChar,
    p => Regexp::Lexer::TokenType::EscapedProp,
    P => Regexp::Lexer::TokenType::EscapedNotProp,
    X => Regexp::Lexer::TokenType::EscapedUnicodeExtendedChar,
    C => Regexp::Lexer::TokenType::EscapedCChar,
    1 => Regexp::Lexer::TokenType::EscapedBackRef,
    2 => Regexp::Lexer::TokenType::EscapedBackRef,
    3 => Regexp::Lexer::TokenType::EscapedBackRef,
    4 => Regexp::Lexer::TokenType::EscapedBackRef,
    5 => Regexp::Lexer::TokenType::EscapedBackRef,
    6 => Regexp::Lexer::TokenType::EscapedBackRef,
    7 => Regexp::Lexer::TokenType::EscapedBackRef,
    8 => Regexp::Lexer::TokenType::EscapedBackRef,
    9 => Regexp::Lexer::TokenType::EscapedBackRef,
    g => Regexp::Lexer::TokenType::EscapedBackRef,
    k => Regexp::Lexer::TokenType::EscapedBackRef,
    K => Regexp::Lexer::TokenType::EscapedKeepStuff,
    v => Regexp::Lexer::TokenType::EscapedVerticalWhitespace,
    V => Regexp::Lexer::TokenType::EscapedNotVerticalWhitespace,
    h => Regexp::Lexer::TokenType::EscapedHorizontalWhitespace,
    H => Regexp::Lexer::TokenType::EscapedNotHorizontalWhitespace,
    R => Regexp::Lexer::TokenType::EscapedLinebreak,
    b => Regexp::Lexer::TokenType::EscapedWordBoundary,
    B => Regexp::Lexer::TokenType::EscapedNotWordBoundary,
    A => Regexp::Lexer::TokenType::EscapedBeginningOfString,
    Z => Regexp::Lexer::TokenType::EscapedEndOfStringBeforeNewline,
    z => Regexp::Lexer::TokenType::EscapedEndOfString,
    G => Regexp::Lexer::TokenType::EscapedPos,
);

my %specialChar = (
    '.'  => Regexp::Lexer::TokenType::MatchAny,
    '|'  => Regexp::Lexer::TokenType::Alternation,
    '('  => Regexp::Lexer::TokenType::LeftParenthesis,
    ')'  => Regexp::Lexer::TokenType::RightParenthesis,
    '['  => Regexp::Lexer::TokenType::LeftBracket,
    ']'  => Regexp::Lexer::TokenType::RightBracket,
    '{'  => Regexp::Lexer::TokenType::LeftBrace,
    '}'  => Regexp::Lexer::TokenType::RightBrace,
    '<'  => Regexp::Lexer::TokenType::LeftAngle,
    '>'  => Regexp::Lexer::TokenType::RightAngle,
    '*'  => Regexp::Lexer::TokenType::Asterisk,
    '+'  => Regexp::Lexer::TokenType::Plus,
    '?'  => Regexp::Lexer::TokenType::Question,
    ','  => Regexp::Lexer::TokenType::Comma,
    '-'  => Regexp::Lexer::TokenType::Minus,
    '$'  => Regexp::Lexer::TokenType::ScalarSigil,
    '@'  => Regexp::Lexer::TokenType::ArraySigil,
    ':'  => Regexp::Lexer::TokenType::Colon,
    '#'  => Regexp::Lexer::TokenType::Sharp,
    '^'  => Regexp::Lexer::TokenType::Cap,
    '='  => Regexp::Lexer::TokenType::Equal,
    '!'  => Regexp::Lexer::TokenType::Exclamation,
    q<'> => Regexp::Lexer::TokenType::SingleQuote,
    q<"> => Regexp::Lexer::TokenType::DoubleQuote,
);

sub tokenize {
    my ($re) = @_;

    if (ref $re ne 'Regexp') {
        croak "Not regexp quoted argument is given";
    }

    # B::cstring() is used to escape backslashes
    my $re_cluster_string = B::cstring($re);

    # to remove double-quotes and parenthesis on leading and trailing
    my $re_str = substr(substr($re_cluster_string, 2), 0, -2);

    $re_str =~ s/\\"/"/g; # for double quote which is converted by B::cstring

    # extract modifiers
    $re_str =~ s/\A[?]([^:]*)://;
    my @modifiers;
    for my $modifier (split //, $1) {
        push @modifiers, $modifier;
    }

    my @chars = split //, $re_str;

    my @tokens;
    my $index = 0;

    my $end_of_line_exists = 0;
    if (defined $chars[-1] && $chars[-1] eq '$') {
        pop @chars;
        $end_of_line_exists = 1;
    }

    if (defined $chars[0] && $chars[0] eq '^') {
        push @tokens, {
            char  => shift @chars,
            index => ++$index,
            type  => Regexp::Lexer::TokenType::BeginningOfLine,
        };
    }

    my $backslashes = 0;
    my $next_c;
    for (my $i = 0; defined(my $c = $chars[$i]); $i++) {
        if ($c eq '\\') {
            if ($backslashes <= 1) {
                $backslashes++;
                next;
            }

            # now status -> '\\\\\\'
            if ($backslashes == 2) {
                $next_c = $chars[++$i];
                if (!defined $next_c || $next_c ne '\\') {
                    croak "Invalid syntax regexp is given"; # fail safe
                }

                push @tokens, {
                    char  => '\\\\',
                    index => ++$index,
                    type  => Regexp::Lexer::TokenType::EscapedCharacter,
                };

                $backslashes = 0;
                next;
            }
        }

        # To support *NOT META* newline character which is in regexp
        if ($backslashes == 1) {
            my $type = Regexp::Lexer::TokenType::Unknown;
            if ($c eq 'n') {
                $type = Regexp::Lexer::TokenType::Newline;
            }
            elsif ($c eq 'r') { # XXX maybe unreachable
                $type = Regexp::Lexer::TokenType::Return;
            }

            push @tokens, {
                char  => '\\' . $c,
                index => ++$index,
                type  => $type,
            };

            $backslashes = 0;
            next;
        }

        if ($backslashes == 2) {
            my $type = $escapedSpecialChar{$c};

            # Determine meaning of \N
            if ($c eq 'N') {
                $type = Regexp::Lexer::TokenType::EscapedCharUnicode;

                $next_c = $chars[$i+1];
                if (!defined $next_c || $next_c ne '{') {
                    $type = Regexp::Lexer::TokenType::EscapedNotNewline;
                }
            }

            push @tokens, {
                char  => '\\' . $c,
                index => ++$index,
                type  => $type || Regexp::Lexer::TokenType::EscapedCharacter,
            };

            $backslashes = 0;
            next;
        }

        push @tokens, {
            char  => $c,
            index => ++$index,
            type  => $specialChar{$c} || Regexp::Lexer::TokenType::Character,
        };

        $backslashes = 0; # for fail safe
    }

    if ($end_of_line_exists) {
        push @tokens, {
            char  => '$',
            index => ++$index,
            type  => Regexp::Lexer::TokenType::EndOfLine,
        };
    }

    return {
        tokens    => \@tokens,
        modifiers => \@modifiers,
    };
}

1;
__END__

=encoding utf-8

=head1 NAME

Regexp::Lexer - Lexer for regular expression of perl

=head1 SYNOPSIS

    use Regexp::Lexer qw(tokenize);
    my $tokens = tokenize(qr{\Ahello\s+world\z}i);

=head1 DESCRIPTION

Regexp::Lexer is a lexer for regular expression of perl.

This module splits the regular expression string to tokens
which has minimum meaning.

=head1 FUNCTIONS

=over 4

=item * C<tokenize($re:Regexp)>

Tokenizes the regular expression.

This function takes an argument as C<Regexp>, namely it must be regexp quoted variable (i.e. C<qr/SOMETHING/>).
If not C<Regexp> argument is given, this function throws exception.

This function returns the result like so;

    {
        tokens => [
            {
                char => '\A',
                index => 1,
                type => {
                    id => 67,
                    name => 'EscapedBeginningOfString',
                },
            },
            {
                char => 'h',
                index => 2,
                type => {
                    id => 1,
                    name => 'Character',
                },
            },
            ...
        ],
        modifiers => ['^', 'i'],
    }

C<tokens> is the tokens list. Information about C<type> of token is located in the L<Regexp::Lexer::TokenType>.

C<modifiers> is the list of modifiers of regular expression. Please see also L<perlre>.

=back

=head1 SEE ALSO

=over 4

=item * L<perlre>

=item * L<perlrebackslash>

=item * L<Regexp::Lexer::TokenType>

=back

=head1 LICENSE

Copyright (C) moznion.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

moznion E<lt>moznion@gmail.comE<gt>

=cut



author/generate_token_type.pl
#!/usr/bin/env perl

# This script generates constants class for token type (Regexp::Lexer::TokenType)
# from `types.yaml`

use strict;
use warnings;
use utf8;
use FindBin;
use YAML::Tiny;

my $yaml = YAML::Tiny->read("$FindBin::Bin/types.yaml")->[0];

my $src = <<'...';
package Regexp::Lexer::TokenType;

# *** DO NOT EDIT THIS FILE DIRECTLY ***
# This file is generated by 'author/generate_token_type.pl'

use strict;
use warnings;
use utf8;

use constant {
...

my $type_value = 1;

my $normal_types = $yaml->{normal};
for my $normal_type (@$normal_types) {
    $src .= <<"...";
    $normal_type => {
        id => $type_value,
        name => '$normal_type',
    },
...
    $type_value++;
};

my $escaped_types = $yaml->{escaped};
for my $escaped_type (@$escaped_types) {
    $src .= <<"...";
    Escaped$escaped_type => {
        id => $type_value,
        name => 'Escaped$escaped_type',
    },
...
    $type_value++;
}

$src .= <<'...';
};

1;
__END__

=encoding utf-8

=head1 NAME

Regexp::Lexer::TokenType - Token types of Regexp::Lexer

=head1 DESCRIPTION

This module provides token types for L<Regexp::Lexer>.

Format of token type is bellow;

    {
        id => <ID of token>,
        name => <name of token>
    }

If you want to identify the token, I highly recommend you to use C<id>.

=head1 TYPES

=over 4

...

for my $normal_type (@$normal_types) {
    $src .= "=item * $normal_type\n\n";
}

for my $escaped_type (@$escaped_types) {
    $src .= "=item * Escaped$escaped_type\n\n";
}

$src .= <<'...';
=back

=head1 LICENSE

Copyright (C) moznion.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

moznion E<lt>moznion@gmail.comE<gt>

=cut

...

my $package_file = "$FindBin::Bin/../lib/Regexp/Lexer/TokenType.pm";

open my $fh, '>', $package_file;
print ${fh} $src;

print "Generated!\n";

__END__



lib/Regexp/Lexer/TokenType.pm
package Regexp::Lexer::TokenType;

# *** DO NOT EDIT THIS FILE DIRECTLY ***
# This file is generated by 'author/generate_token_type.pl'

use strict;
use warnings;
use utf8;

use constant {
    Character => {
        id => 1,
        name => 'Character',
    },
    BeginningOfLine => {
        id => 2,
        name => 'BeginningOfLine',
    },
    EndOfLine => {
        id => 3,
        name => 'EndOfLine',
    },
    MatchAny => {
        id => 4,
        name => 'MatchAny',
    },
    Alternation => {
        id => 5,
        name => 'Alternation',
    },
    LeftParenthesis => {
        id => 6,
        name => 'LeftParenthesis',
    },
    RightParenthesis => {
        id => 7,
        name => 'RightParenthesis',
    },
    LeftBracket => {
        id => 8,
        name => 'LeftBracket',
    },
    RightBracket => {
        id => 9,
        name => 'RightBracket',
    },
    LeftBrace => {
        id => 10,
        name => 'LeftBrace',
    },
    RightBrace => {
        id => 11,
        name => 'RightBrace',
    },
    LeftAngle => {
        id => 12,
        name => 'LeftAngle',
    },
    RightAngle => {
        id => 13,
        name => 'RightAngle',
    },
    Equal => {
        id => 14,
        name => 'Equal',
    },
    Plus => {
        id => 15,
        name => 'Plus',
    },
    Minus => {
        id => 16,
        name => 'Minus',
    },
    Asterisk => {
        id => 17,
        name => 'Asterisk',
    },
    Sharp => {
        id => 18,
        name => 'Sharp',
    },
    Cap => {
        id => 19,
        name => 'Cap',
    },
    Comma => {
        id => 20,
        name => 'Comma',
    },
    Colon => {
        id => 21,
        name => 'Colon',
    },
    Exclamation => {
        id => 22,
        name => 'Exclamation',
    },
    Question => {
        id => 23,
        name => 'Question',
    },
    ArraySigil => {
        id => 24,
        name => 'ArraySigil',
    },
    ScalarSigil => {
        id => 25,
        name => 'ScalarSigil',
    },
    SingleQuote => {
        id => 26,
        name => 'SingleQuote',
    },
    DoubleQuote => {
        id => 27,
        name => 'DoubleQuote',
    },
    Newline => {
        id => 28,
        name => 'Newline',
    },
    Return => {
        id => 29,
        name => 'Return',
    },
    Unknown => {
        id => 30,
        name => 'Unknown',
    },
    EscapedCharacter => {
        id => 31,
        name => 'EscapedCharacter',
    },
    EscapedTab => {
        id => 32,
        name => 'EscapedTab',
    },
    EscapedNewline => {
        id => 33,
        name => 'EscapedNewline',
    },
    EscapedReturn => {
        id => 34,
        name => 'EscapedReturn',
    },
    EscapedFormFeed => {
        id => 35,
        name => 'EscapedFormFeed',
    },
    EscapedAlarm => {
        id => 36,
        name => 'EscapedAlarm',
    },
    EscapedEscape => {
        id => 37,
        name => 'EscapedEscape',
    },
    EscapedControlChar => {
        id => 38,
        name => 'EscapedControlChar',
    },
    EscapedCharHex => {
        id => 39,
        name => 'EscapedCharHex',
    },
    EscapedCharUnicode => {
        id => 40,
        name => 'EscapedCharUnicode',
    },
    EscapedCharOct => {
        id => 41,
        name => 'EscapedCharOct',
    },
    EscapedLowerNext => {
        id => 42,
        name => 'EscapedLowerNext',
    },
    EscapedUpperNext => {
        id => 43,
        name => 'EscapedUpperNext',
    },
    EscapedLowerUntil => {
        id => 44,
        name => 'EscapedLowerUntil',
    },
    EscapedUpperUntil => {
        id => 45,
        name => 'EscapedUpperUntil',
    },
    EscapedQuoteMetaUntil => {
        id => 46,
        name => 'EscapedQuoteMetaUntil',
    },
    EscapedEnd => {
        id => 47,
        name => 'EscapedEnd',
    },
    EscapedWordChar => {
        id => 48,
        name => 'EscapedWordChar',
    },
    EscapedNotWordChar => {
        id => 49,
        name => 'EscapedNotWordChar',
    },
    EscapedWhiteSpaceChar => {
        id => 50,
        name => 'EscapedWhiteSpaceChar',
    },
    EscapedNotWhiteSpaceChar => {
        id => 51,
        name => 'EscapedNotWhiteSpaceChar',
    },
    EscapedDigitChar => {
        id => 52,
        name => 'EscapedDigitChar',
    },
    EscapedNotDigitChar => {
        id => 53,
        name => 'EscapedNotDigitChar',
    },
    EscapedProp => {
        id => 54,
        name => 'EscapedProp',
    },
    EscapedNotProp => {
        id => 55,
        name => 'EscapedNotProp',
    },
    EscapedUnicodeExtendedChar => {
        id => 56,
        name => 'EscapedUnicodeExtendedChar',
    },
    EscapedCChar => {
        id => 57,
        name => 'EscapedCChar',
    },
    EscapedBackRef => {
        id => 58,
        name => 'EscapedBackRef',
    },
    EscapedKeepStuff => {
        id => 59,
        name => 'EscapedKeepStuff',
    },
    EscapedNotNewline => {
        id => 60,
        name => 'EscapedNotNewline',
    },
    EscapedVerticalWhitespace => {
        id => 61,
        name => 'EscapedVerticalWhitespace',
    },
    EscapedNotVerticalWhitespace => {
        id => 62,
        name => 'EscapedNotVerticalWhitespace',
    },
    EscapedHorizontalWhitespace => {
        id => 63,
        name => 'EscapedHorizontalWhitespace',
    },
    EscapedNotHorizontalWhitespace => {
        id => 64,
        name => 'EscapedNotHorizontalWhitespace',
    },
    EscapedLinebreak => {
        id => 65,
        name => 'EscapedLinebreak',
    },
    EscapedWordBoundary => {
        id => 66,
        name => 'EscapedWordBoundary',
    },
    EscapedNotWordBoundary => {
        id => 67,
        name => 'EscapedNotWordBoundary',
    },
    EscapedBeginningOfString => {
        id => 68,
        name => 'EscapedBeginningOfString',
    },
    EscapedEndOfStringBeforeNewline => {
        id => 69,
        name => 'EscapedEndOfStringBeforeNewline',
    },
    EscapedEndOfString => {
        id => 70,
        name => 'EscapedEndOfString',
    },
    EscapedPos => {
        id => 71,
        name => 'EscapedPos',
    },
    EscapedFoldcase => {
        id => 72,
        name => 'EscapedFoldcase',
    },
};

1;
__END__

=encoding utf-8

=head1 NAME

Regexp::Lexer::TokenType - Token types of Regexp::Lexer

=head1 DESCRIPTION

This module provides token types for L<Regexp::Lexer>.

Format of token type is bellow;

    {
        id => <ID of token>,
        name => <name of token>
    }

If you want to identify the token, I highly recommend you to use C<id>.

=head1 TYPES

=over 4

=item * Character

=item * BeginningOfLine

=item * EndOfLine

=item * MatchAny

=item * Alternation

=item * LeftParenthesis

=item * RightParenthesis

=item * LeftBracket

=item * RightBracket

=item * LeftBrace

=item * RightBrace

=item * LeftAngle

=item * RightAngle

=item * Equal

=item * Plus

=item * Minus

=item * Asterisk

=item * Sharp

=item * Cap

=item * Comma

=item * Colon

=item * Exclamation

=item * Question

=item * ArraySigil

=item * ScalarSigil

=item * SingleQuote

=item * DoubleQuote

=item * Newline

=item * Return

=item * Unknown

=item * EscapedCharacter

=item * EscapedTab

=item * EscapedNewline

=item * EscapedReturn

=item * EscapedFormFeed

=item * EscapedAlarm

=item * EscapedEscape

=item * EscapedControlChar

=item * EscapedCharHex

=item * EscapedCharUnicode

=item * EscapedCharOct

=item * EscapedLowerNext

=item * EscapedUpperNext

=item * EscapedLowerUntil

=item * EscapedUpperUntil

=item * EscapedQuoteMetaUntil

=item * EscapedEnd

=item * EscapedWordChar

=item * EscapedNotWordChar

=item * EscapedWhiteSpaceChar

=item * EscapedNotWhiteSpaceChar

=item * EscapedDigitChar

=item * EscapedNotDigitChar

=item * EscapedProp

=item * EscapedNotProp

=item * EscapedUnicodeExtendedChar

=item * EscapedCChar

=item * EscapedBackRef

=item * EscapedKeepStuff

=item * EscapedNotNewline

=item * EscapedVerticalWhitespace

=item * EscapedNotVerticalWhitespace

=item * EscapedHorizontalWhitespace

=item * EscapedNotHorizontalWhitespace

=item * EscapedLinebreak

=item * EscapedWordBoundary

=item * EscapedNotWordBoundary

=item * EscapedBeginningOfString

=item * EscapedEndOfStringBeforeNewline

=item * EscapedEndOfString

=item * EscapedPos

=item * EscapedFoldcase

=back

=head1 LICENSE

Copyright (C) moznion.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

moznion E<lt>moznion@gmail.comE<gt>

=cut

